[flang] optimize array function calls using hlfir.eval_in_mem

jeanPerier opened the pull request:
This patch encapsulate array function call lowering into hlfir.eval_in_mem and allows directly evaluating the call into the LHS when possible.

The conditions are: LHS is contiguous, not accessed inside the function, it is not a whole allocatable, and the function results needs not to be finalized. All these conditions are tested in the previous [hlfir.eval_in_mem optimization patch](https://github.com/llvm/llvm-project/pull/118069) that is leveraging [the extension of getModRef to handle function calls](https://github.com/llvm/llvm-project/pull/117164)).

This yields a 25% speed-up on polyhedron channel2 benchmark (from 1min to 45s measured on an X86-64 Zen 2).

jeanPerier committed:
[flang][hlfir] add hlfir.eval_in_mem operation

jeanPerier committed:
[flang][hlfir] optimize hlfir.eval_in_mem bufferization

jeanPerier committed:
[flang] optimize array function calls using hlfir.eval_in_mem

jeanPerier commented:
Note that this is a "classic" Fortran compiler optimization, you can verify that at least all of gfortran, ifx, and nvfortran are doing it on something very easy like:

```
subroutine test
interface
 function foo()
    real :: foo(100)
 end function
end interface
   real x(100)
   x = foo()
   call bar(x)
end subroutine
```

tblah commented on flang/include/flang/Lower/ConvertCall.h:
```diff
@@ -24,6 +24,10 @@
 
 namespace Fortran::lower {
 
+struct LoweredResult {
+  std::variant<fir::ExtendedValue, hlfir::EntityWithAttributes> result;
+};
```
nit: Could this be simplified to
```c++
using LoweredResult = std::variant<fir::ExtendedValue, hlfir::EntityWithAttributes> result;
```

This way callers wouldn't have to add a `.result`.

jeanPerier replied:
Updated.

tblah commented on flang/test/HLFIR/order_assignments/where-scheduling.f90:
```diff
@@ -134,7 +134,7 @@ end function f
 !CHECK-NEXT: run 2 evaluate: where/region_assign1
 !CHECK-LABEL: ------------ scheduling where in _QPonly_once ------------
-!CHECK-NEXT: unknown effect: %{{[0-9]+}} = llvm.intr.stacksave : !llvm.ptr
+!CHECK-NEXT: unknown effect: %11 = fir.call @_QPcall_me_only_once() fastmath<contract> : () -> !fir.array<10x!fir.logical<4>>
```
Why did this change?

tblah replied:
Thanks for the explanation

jeanPerier committed:
PR118069 comment: mayReadOrWrite to getModRef

jeanPerier committed:
Merge branch 'main' into users/jperier/opt_eval_in_mem

jeanPerier committed:
Merge branch 'opt_eval_in_mem' into array_call_opt

tblah reviewed (approved):
LGTM with the one nit

jeanPerier committed:
PR118070 comments: simplify LoweredResult

jeanPerier committed:
remove new over-assertive assert

I added an assert that prevents ExprType to be "trivial types" (i32, f32, ...)
but that actually breaks existing lowering code that generate as_expr + associate
patterns to create temporaries:

https://github.com/llvm/llvm-project/blob/03b5f8f0f0d10c412842ed04b90e2217cf071218/flang/lib/Lower/ConvertCall.cpp#L1271

```
subroutine trivial_as_expr(i)
  integer, optional :: i
  interface
    subroutine foo(j)
      integer, optional, value :: j
    end subroutine
  end interface
  call foo(i)
end subroutine
```

This is unrelated to this patch, so leave it like this (it is not even a
functional problem).

jeanPerier commented:
Thanks for all the reviews Tom!
